doniclient/v1/client.py

```diff
@@ -50,11 +50,26 @@ def create(self, json, **kwargs):
         except json.JSONDecodeError:
             return resp
 
-    def delete(self, uuid):
-        return self.adapter.delete(f"/v1/hardware/{uuid}/")
+    def delete(self, name_or_uuid):
+        hardware_list = self.list()
+        matching_hardware = [h for h in hardware_list if h['name'] == name_or_uuid]
+        hardware_uuid = name_or_uuid
+        if matching_hardware:
+            hardware_uuid = matching_hardware[0]['uuid']
+        return self.adapter.delete(f"/v1/hardware/{hardware_uuid}/")
+        
 
     def sync(self, uuid):
         return self.adapter.post(f"/v1/hardware/{uuid}/sync")
 
-    def update(self, uuid, json):
-        return self.adapter.patch(f"/v1/hardware/{uuid}/", json=json).json()
+    def update(self, name_or_uuid, json):
+        hardware_list = self.list()
+        matching_hardware = [h for h in hardware_list if h['name'] == name_or_uuid]
+        hardware_uuid = name_or_uuid
+        if matching_hardware:
+            hardware_uuid = matching_hardware[0]['uuid']
+        resp = self.adapter.patch(f"/v1/hardware/{hardware_uuid}/", json=json)
+        try:
+            return resp.json()
+        except json.JSONDecodeError:
+            return resp
```

Comment:
I would prefer if we kept the methods in `/v1/client.py` a relatively 1-1 match to the REST API, as other python code may import this library to use that.

Instead, we can add the extra logic to the CLI methods under `osc`.

For example, replacing:
`hw_client.delete(parsed_args.uuid)`
with
```
# find_exactly_one should throw an error if more than one match is found.
hardware_instance = find_exactly_one(hw_client, name_or_uuid)
hw_client.delete(hardware_instance.uuid)
```